PR: https://github.com/odoo/odoo/pull/241780

From: 6057e25f8c1286e423a9d9686307ae3be5808c2c
From: karygauss03
Date: 2026-03-13 10:11:04

Breaking data model changes scores: del:5 + add:0, change matches:
-    contract_type_id = fields.Many2one(readonly=False, related='version_id.contract_type_id', inherited=True, groups="hr.group_hr_manager")
-    has_country_contract_type = fields.Boolean(
-    contract_type_id = fields.Many2one('hr.contract.type', string='Contract Type', tracking=True)
-    employee_type = fields.Selection([
-    contract_type_id = fields.Many2one('hr.contract.type', "Contract Type", tracking=1,

Total Changes: 103

[IMP] hr: Remove employee_type and use contract_type_id

In this commit, we removed employee_type from hr_version and hr_employee
and we change its usage to contract_type_id.
Also, we renamed the field Contract Type to Employee Type.

task-5430894

closes odoo/odoo#241780

Related: odoo/enterprise#103118
Related: odoo/upgrade#9294
Signed-off-by: Yannick Tivisse (yti) <yti@odoo.com>

================================= pseudo patch: =================================

--- a/addons/hr/models/__init__.py
+++ b/addons/hr/models/__init__.py
@@ -3,7 +3,7 @@
 from . import hr_payroll_structure_type
 from . import hr_job
 from . import hr_version
-from . import hr_contract_type
+from . import hr_employee_type
 from . import mail_activity
 from . import hr_employee
 from . import hr_mixin

--- a/addons/hr/models/hr_contract_type.py
+++ b/None
@@ -1,21 +0,0 @@
-# -*- coding: utf-8 -*-
-# Part of Odoo. See LICENSE file for full copyright and licensing details.
-
-from odoo import api, fields, models
-
-
-class HrContractType(models.Model):
-    _name = 'hr.contract.type'
-    _description = 'Contract Type'
-    _order = 'name'
-
-    name = fields.Char(required=True, translate=True)
-    code = fields.Char(compute='_compute_code', store=True, readonly=False)
-    country_id = fields.Many2one('res.country', domain=lambda self: [('id', 'in', self.env.companies.country_id.ids)])
-
-    @api.depends('name')
-    def _compute_code(self):
-        for contract_type in self:
-            if contract_type.code:
-                continue
-            contract_type.code = contract_type.name

--- a/addons/hr/models/hr_employee.py
+++ b/addons/hr/models/hr_employee.py
@@ -221,6 +221,7 @@ class HrEmployee(models.Model):
                                     help="The date of the first contract of the employee in the company.")
     contract_date_start = fields.Date(readonly=False, related="version_id.contract_date_start", inherited=True, groups="hr.group_hr_manager")
     contract_date_end = fields.Date(readonly=False, related="version_id.contract_date_end", inherited=True, groups="hr.group_hr_manager")
+    fixed_term = fields.Boolean(readonly=False, related="version_id.fixed_term", inherited=True, groups="hr.group_hr_manager")
     trial_date_end = fields.Date(readonly=False, related="version_id.trial_date_end", inherited=True, groups="hr.group_hr_manager")
     date_start = fields.Date(related='version_id.date_start', inherited=True, groups="hr.group_hr_manager")
     date_end = fields.Date(related='version_id.date_end', inherited=True, groups="hr.group_hr_manager")
@@ -229,7 +230,7 @@ class HrEmployee(models.Model):
     is_future = fields.Boolean(related='version_id.is_future', inherited=True, groups="hr.group_hr_manager")
     is_in_contract = fields.Boolean(related='version_id.is_in_contract', inherited=True, groups="hr.group_hr_manager")
     structure_type_id = fields.Many2one(readonly=False, related='version_id.structure_type_id', inherited=True, groups="hr.group_hr_manager")
-    contract_type_id = fields.Many2one(readonly=False, related='version_id.contract_type_id', inherited=True, groups="hr.group_hr_manager")
+    employee_type_id = fields.Many2one(readonly=False, related='version_id.employee_type_id', inherited=True, groups="hr.group_hr_manager")
     hourly_cost = fields.Monetary('Hourly Cost', groups="hr.group_hr_user", tracking=True)
     nationality_country_code = fields.Char(
         string='Nationality',
@@ -351,8 +352,8 @@ class HrEmployee(models.Model):
         'A user cannot be linked to multiple employees in the same company.',
     )
 
-    has_country_contract_type = fields.Boolean(
-        compute='_compute_has_country_contract_type',
+    has_country_employee_type = fields.Boolean(
+        compute='_compute_has_country_employee_type',
         groups="hr.group_hr_user",
     )
 
@@ -719,14 +720,14 @@ class HrEmployee(models.Model):
                 employee.current_version_id = new_current_version
 
     @api.depends('company_country_id')
-    def _compute_has_country_contract_type(self):
-        count_contract_type_by_country = dict(self.env['hr.contract.type']._read_group(
+    def _compute_has_country_employee_type(self):
+        count_contract_type_by_country = dict(self.env['hr.employee.type']._read_group(
             domain=[],
             groupby=['country_id'],
             aggregates=['__count']
         ))
         for employee in self:
-            employee.has_country_contract_type = bool(
+            employee.has_country_employee_type = bool(
                 count_contract_type_by_country.get(employee.company_country_id)
             )
 

--- a/None
+++ b/addons/hr/models/hr_employee_type.py
@@ -0,0 +1,49 @@
+from odoo import api, fields, models
+
+
+class HrEmployeeType(models.Model):
+    _name = 'hr.employee.type'
+    _description = 'Employee Type'
+    _order = 'name'
+
+    name = fields.Char(required=True, translate=True)
+    code = fields.Char(compute='_compute_code', store=True, readonly=False)
+    country_id = fields.Many2one('res.country', domain=lambda self: [('id', 'in', self.env.companies.country_id.ids)])
+    employees_count = fields.Integer(compute='_compute_employee_count', string='Employees')
+
+    @api.depends('name')
+    def _compute_code(self):
+        for contract_type in self:
+            if contract_type.code:
+                continue
+            contract_type.code = contract_type.name
+
+    def _compute_employee_count(self):
+        employee_count_by_employee_type = dict(self.env['hr.employee']._read_group(
+            domain=[('employee_type_id', 'in', self.ids)],
+            groupby=['employee_type_id'],
+            aggregates=['id:count'],
+        ))
+        for employee_type in self:
+            employee_type.employees_count = employee_count_by_employee_type.get(employee_type, 0)
+
+    def action_open_employees(self):
+        self.ensure_one()
+        employees = self.env['hr.employee'].search([('employee_type_id', 'in', self.ids)])
+        if len(employees) == 1:
+            return {
+                'name': self.env._('Employee'),
+                'type': 'ir.actions.act_window',
+                'res_model': 'hr.employee',
+                'res_id': employees.id,
+                'view_mode': 'form',
+                'views': [(False, 'form')],
+            }
+        return {
+            'name': self.env._('Related Employees'),
+            'type': 'ir.actions.act_window',
+            'res_model': 'hr.employee',
+            'view_mode': 'kanban, form',
+            'views': [(False, 'kanban'), (False, 'form')],
+            'domain': [('id', 'in', employees.ids)],
+        }

--- a/addons/hr/models/hr_job.py
+++ b/addons/hr/models/hr_job.py
@@ -42,7 +42,7 @@ class HrJob(models.Model):
     )
     department_id = fields.Many2one('hr.department', string='Department', check_company=True, tracking=True, index='btree_not_null')
     company_id = fields.Many2one('res.company', string='Company', default=lambda self: self.env.company, tracking=True)
-    contract_type_id = fields.Many2one('hr.contract.type', string='Contract Type', tracking=True)
+    employee_type_id = fields.Many2one('hr.employee.type', string='Employee Type', tracking=True)
 
     _name_company_uniq = models.Constraint(
         'unique(name, company_id, department_id)',

--- a/addons/hr/models/hr_version.py
+++ b/addons/hr/models/hr_version.py
@@ -106,17 +106,6 @@ class HrVersion(models.Model):
     children = fields.Integer(string='Dependent Children', groups="hr.group_hr_user", tracking=1)
 
     # Work Information
-    employee_type = fields.Selection([
-            ('employee', 'Employee'),
-            ('worker', 'Worker'),
-            ('student', 'Student'),
-            ('trainee', 'Trainee'),
-            ('contractor', 'Contractor'),
-            ('freelance', 'Freelancer'),
-            ('statutory', 'Statutory'),
-            ('apprenticeship', 'Apprenticeship'),
-            ('company_executive', 'Company Executive')
-        ], string='Employee Type', default='employee', required=True, groups="hr.group_hr_user", tracking=1)
     department_id = fields.Many2one('hr.department', check_company=True, tracking=1, index=True)
     member_of_department = fields.Boolean("Member of department", compute='_compute_part_of_department', search='_search_part_of_department',
         help="Whether the employee is a member of the active user's department or one of it's child department.")
@@ -161,6 +150,7 @@ class HrVersion(models.Model):
     contract_date_end = fields.Date(
         'Contract End Date', tracking=1, help="End date of the contract (if it's a fixed-term contract).",
         groups="hr.group_hr_manager")
+    fixed_term = fields.Boolean('Fixed Term', tracking=1, groups='hr.group_hr_manager')
     trial_date_end = fields.Date('End of Trial Period', help="End date of the trial period (if there is one).",
                                  groups="hr.group_hr_manager", tracking=1)
     date_start = fields.Date(compute='_compute_dates', groups="hr.group_hr_manager", search="_search_start_date")
@@ -185,7 +175,7 @@ class HrVersion(models.Model):
     company_country_id = fields.Many2one('res.country', string="Company country",
                                          related='company_id.country_id', readonly=True)
     country_code = fields.Char(related='company_country_id.code', depends=['company_country_id'], readonly=True)
-    contract_type_id = fields.Many2one('hr.contract.type', "Contract Type", tracking=1,
+    employee_type_id = fields.Many2one('hr.employee.type', "Employee Type", tracking=1,
                                        groups="hr.group_hr_manager")
     additional_note = fields.Text(string='Additional Note', groups="hr.group_hr_user", tracking=1)
 
@@ -450,7 +440,7 @@ class HrVersion(models.Model):
     def _get_whitelist_fields_from_template(self):
         # Add here any field that you want to copy from a contract template
         # Those fields should have tracking=1 in hr.version to see the change
-        return ['job_id', 'department_id', 'contract_type_id', 'structure_type_id', 'wage', 'resource_calendar_id', 'hr_responsible_id']
+        return ['job_id', 'department_id', 'employee_type_id', 'structure_type_id', 'wage', 'resource_calendar_id', 'hr_responsible_id']
 
     def get_values_from_contract_template(self, contract_template_id):
         if not contract_template_id:
